[22.05] Fix: skip setting metadata on deferred datasets when changing datatype in bulk

davelopez opened the pull request:
Fixes #14096 

When running the *Change datatype* operation in bulk, changing the data type on a deferred dataset seems ok as long as we don't alter the state of the dataset. Generating the metadata doesn't work because the dataset is not materialized so we skip it.

I'm not completely sure this is the correct fix, so suggestions are greatly welcome.

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  1. Create one or more deferred dataset
  2. Run the `Change datatype` operation from the selection menu on them
  3. The state of the dataset should still be deferred

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

davelopez committed:
Add API test to skip deferred set_metadata in bulk

davelopez committed:
Skip set metadata on deferred datasets

Using celery task

davelopez committed:
Skip setting the state for deferred dataset

When changing the datatype in bulk of a deferred dataset we shouldn't alter the state of the dataset or it won't be treated as deferred anymore.

davelopez commented on lib/galaxy_test/api/test_history_contents.py:
```diff
@@ -1266,6 +1266,38 @@ def test_bulk_datatype_change(self):
+            self._assert_bulk_success(bulk_operation_result, expected_success_count=1)
+
+            # Wait for celery tasks to finish
+            time.sleep(2)  # I don't like this at all, is there another way to wait for celery here?
```
Is there a better way to wait for celery to finish here? Unfortunately waiting for a history change doesn't help.

mvdbeek replied:
~~You're not running a celery task, so ... I don't think you need to wait ? Just assert that the dataset is not in `setting_metadata` should be suffcient.~~

mvdbeek replied:
So my first comment is nonsense, but changing the datatype without setting metadata (so what you do for deferred datasets) doesn't have to happen in a celery task, there's very little I/O or CPU involved, I think it's fine to do that during the request for now.

davelopez replied:
Is c58e3f4 more or less what you had in mind?

mvdbeek replied:
Yes, perfect.

davelopez commented on lib/galaxy_test/api/test_history_contents.py:
```diff
@@ -1266,6 +1266,38 @@ def test_bulk_datatype_change(self):
+                assert item["extension"] == "txt"
+                assert item["data_type"] == "galaxy.datatypes.data.Text"
+                # Should keep or discard old metadata?
+                assert "metadata_columns" not in item
```
I'm not sure in what state should be the previous metadata (when creating the dataset) after changing the data type. It seems to be dropped which kind of makes sense?

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1621,8 +1621,9 @@ def _change_datatype(
+            if not item.has_deferred_data:
+                item.dataset.state = item.dataset.states.SETTING_METADATA
+                trans.sa_session.flush()
             change_datatype.delay(dataset_id=item.id, datatype=params.datatype)
```
```suggestion
                trans.sa_session.flush()
                change_datatype.delay(dataset_id=item.id, datatype=params.datatype)
```
My comment below only makes sense with that change ...

mvdbeek replied:
sorry, I should take more time to read this all of this.

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1621,8 +1621,9 @@ def _change_datatype(
-            trans.sa_session.flush()
+            if not item.has_deferred_data:
+                item.dataset.state = item.dataset.states.SETTING_METADATA
+                trans.sa_session.flush()
```
```suggestion
            item.dataset.state = item.dataset.states.SETTING_METADATA
            trans.sa_session.flush()
```
I would still set the state to SETTING_METADATA, so no-one can start a job with this dataset as input. Thinking about it, maybe `ensure_can_set_metadata` (or just `ok_to_edit_metadata` without the exception) should go into the task as well (See the second paragraph in https://docs.celeryq.dev/en/latest/userguide/tasks.html#state, `“asserting the world is the responsibility of the task”.`).

davelopez replied:
Ok, I've done some changes for this in 739b2e8.

davelopez committed:
Ensure `world` state before running some task

- Add parameter to ensure_can_change_datatype and ensure_can_set_metadata to return a boolean instead of raising exceptions
- Check that we can perform the actions in the actual change_datatype and set_metadata celery tasks

davelopez committed:
Change datatype of deferred directly in request

Instead of spinning a celery task we just set the new datatype and skip the sniffing if `auto-detect` was requested.
